### PR TITLE
Don't build the portable runtime by default

### PR DESCRIPTION
**Pull request by crummel**
Per https://github.com/dotnet/source-build/issues/1529, we believe this is the more common case.  Users who wish to generate a tarball on banana-linux and build from the tarball on apple-linux can still re-enable this behavior.

### CODE REVIEW

File: dir.props [hunk 1]
```diff
@@ -14,6 +14,7 @@
     <PortableBuild Condition="'$(PortableBuild)' == ''">false</PortableBuild>
     <UseSystemLibraries Condition="'$(UseSystemLibraries)' == '' AND '$(PortableBuild)' != 'true'">true</UseSystemLibraries>
     <UseStableVersions Condition="'$(UseStableVersions)' == ''">true</UseStableVersions>
+    <SkipPortableRuntimeBuild Condition="'(SkipPortableRuntimeBuild)' == ''">true</SkipPortableRuntimeBuild>
```
Comment (dagood) [inline dir.props]: Can we instead change the variable to `PortableRuntimeBuild` to avoid a double negative build arg and use the slightly more intuitive `false` default?

File: dir.props [hunk 2]
```diff
@@ -14,6 +14,7 @@
     <PortableBuild Condition="'$(PortableBuild)' == ''">false</PortableBuild>
     <UseSystemLibraries Condition="'$(UseSystemLibraries)' == '' AND '$(PortableBuild)' != 'true'">true</UseSystemLibraries>
     <UseStableVersions Condition="'$(UseStableVersions)' == ''">true</UseStableVersions>
+    <BuildPortableRuntime Condition="'(BuildPortableRuntime)' == ''">false</BuildPortableRuntime>
```
Comment (dagood) [inline dir.props]: Should we turn this on in any of our CI/PR validation legs?

Comment (crummel) [reply dir.props]: I think it's probably not necessary - we have a couple portable-only builds so the only thing we could really catch is if building both portable and non-portable breaks something, which doesn't seem very likely to me.  What do you think?

Comment (dagood) [reply dir.props]: Does this mean there's no known users of this feature now? Not that we have to delete it, but without validation it'll surely break at some point.

Comment (omajid) [reply dir.props]: @dagood As I understand it, we are using this (or, well, we were using this) for the Feb release on RHEL 8. I need to look at bootstrapping RHEL 8, but that's going to be harder to justify internally than a normal version-bump-bugfix release. So I suggest we should keep it until source-build downstreams can migrate away.

Comment (dagood) [reply dir.props]: @omajid Would it have been reasonable to use a nonportable SDK on RHEL7 to build a portable SDK, then use that "one-off" portable SDK as the N-1 to build a nonportable SDK on RHEL8?

That seems to me to accomplish essentially the same thing as this feature, but easier to understand. However, it does mean building a whole portable SDK rather than just some portable previously-source-built bits.

Another factor is whether it's possible to do that... I don't know of any issues preventing it though.

I'm wondering if this situation (removing libs) is common enough with major upgrades that distro maintainers would like a very easy answer here. Would distro maintainers in general prefer:
* to carry along a normally unused portable build through their build chain (this flag) to be able to use it for a new major distro version at any time with no startup cost? (But with potentially significant per-build cost?)
* do a one-off portable build like the above? (We should consider what validation we might need to be able to produce a portable build using a nonportable N-1.)
* rebootstrap? (Surely not, but seems it would be an option.)

Comment (omajid) [reply dir.props]: > Would it have been reasonable to use a nonportable SDK on RHEL7 to build a portable SDK, then use that "one-off" portable SDK as the N-1 to build a nonportable SDK on RHEL8?

Yes, this sounds very sensible approach, in general.

(The only trouble is that doing something like this in the middle of a RHEL 8 release is quite a bit harder to do. If we are just doing RHEL 8 for the first time - like we just did Fedora - this would have been a very reasonable approach.)

> That seems to me to accomplish essentially the same thing as this feature, but easier to understand. However, it does mean building a whole portable SDK rather than just some portable previously-source-built bits.

This would be one-off, so doesn't seem that bad.

> - do a one-off portable build like the above? (We should consider what validation we might need to be able to produce a portable build using a nonportable N-1.)
> - rebootstrap? (Surely not, but seems it would be an option.)

Can you help me understand the distinction between the two here? They both seem like starting with a prebuilt SDK: "one-off portable" is source-built on an older distro and "rebootstrap" is the source-built built by Microsoft (or source-build upstream)?

Comment (dagood) [reply dir.props]: > > * do a one-off portable build like the above? (We should consider what validation we might need to be able to produce a portable build using a nonportable N-1.)
> > * rebootstrap? (Surely not, but seems it would be an option.)
> 
> Can you help me understand the distinction between the two here? They both seem like starting with a prebuilt SDK: "one-off portable" is source-built on an older distro and "rebootstrap" is the source-built built by Microsoft (or source-build upstream)?

Well, my assumption here is that the portable SDK built using RHEL 7's nonportable SDK is still considered buildable from source when you use it in RHEL 8, making it an N-1 flow (single build), not a situation that requires re-bootstrapping (multiple builds) to prove it's buildable from source.

I suppose this assumption doesn't make sense because RHEL 7 will go out of support sooner than RHEL 8, and once RHEL 7 goes out of support you can't truly know the .NET Core on RHEL 8 is buildable on RHEL 8. But that makes me question the whole scenario... is it possible to use N-1 flow across major version boundaries at all? In what situations can you use RHEL 7 bits to build a package for RHEL 8? (Maybe I'm missing the forest for the trees with this policy?)

Comment (omajid) [reply dir.props]: Would it help if I describe how distributions like Fedora (and thus, RHEL) do updates?

- For bootstrapping, we start with a prebuilt binary. We dont really care who built it (whether it's N-1 from source-build, or a binary downloaded from a project's website, as long as it's under an Open Source license). We then put that in the "sources" and then do a build using that as the "compiler" or "runtime". Once we have a package built in the distro, we use that package for future builds. The "initial" binary is never used again. It may not even be available, ever, outside the special build system.

- For updates, such as going from 3.1.101 to 3.1.102, we start with the existing package in the distro. On Fedora, we install the `dotnet-sdk-3.1` (= 3.1.101) package in the distro, then use that to compile 3.1.102. Now the distro has 3.1.102.

- Multiple versions of the distribution are disconnected. Fedora 31, for example, wont see packages added/changed in Fedora 30 (aside from when Fedora 31 was initially forked from Fedora 30). Or vice-versa. Same goes for RHEL 7 and RHEL 8.

> Well, my assumption here is that the portable SDK built using RHEL 7's nonportable SDK is still considered buildable from source when you use it in RHEL 8, making it an N-1 flow (single build), not a situation that requires re-bootstrapping (multiple builds) to prove it's buildable from source.

I guess what I am trying to say is that using something built on RHEL 7 on RHEL 8 is just as much of a bootstrap from the distro's point of view as using a prebuilt binary built elsewhere.

There's a gotcha here: 3.1.100 didn't have a working bootstrap process, so what we did in RHEL 7/RHEL 8 is using the prebuilts as published by source-build. They dont use the bootstrap process and adding them right now is not straight forward thanks to how RHEL is generally locked down for "features".

Comment (dagood) [reply dir.props]: It does, thanks. My struggle here has been trying to reconcile why building nonportable+portable was useful, and made (it turns out a lot of) assumptions when I heard it was being used for some bootstrapping problem in RHEL 8. I'm curious what this actually solved, but I suppose that's just a historical note. I'll get caught up with the folks over here to avoid wasting time. 😛

> > In what situations can you use RHEL 7 bits to build a package for RHEL 8? [...]
> 
> We can't do this in Fedora. [...] The only time a package from Fedora N ends up in Fedora N+1 is when Fedora N+1 is initially forked from Fedora N.

I still don't quite get this. It sounds like you *can* use the SDK package from Fedora N to build .NET Core for Fedora N+1.
Is Fedora N+1 forking from Fedora N a significant edge case?
Or does "ending up in" still mean it would have to be treated as a prebuilt?

> I guess what I am trying to say is that using something built on RHEL 7 on RHEL 8 is just as much of a bootstrap from the distro's point of view as using a prebuilt binary built elsewhere.

I got wrapped up in the importance of a line of continuity, and what you're describing is purer than what I was thinking. Appreciate it. :+1:

Comment (omajid) [reply dir.props]: > I still don't quite get this. It sounds like you can use the SDK package from Fedora N to build .NET Core for Fedora N+1.
> Is Fedora N+1 forking from Fedora N a significant edge case?

Lets assume we are on Fedora 35 right now. And it has a working .NET Core SDK package.

When it's time to create Fedora 36 - the distribution. Fedora folks "fork" (or copy) the packages from Fedora 35 to create Fedora 36. So Fedora 36 starts with the same working package (the exact same bits) that was present in Fedora 35.

But as soon as that initial fork is done, Fedora 35 and Fedora 36 are completely disconnected. I can't use anything from Fedora 35 in Fedora 36 anymore. If I now build a .NET Core SDK in Fedora 36, it is using that package that was originally copied from Fedora 35.

In a way, we use a package (that was originally built on Fedora 35) to build the package on Fedora 36. But that's a one-time event. We can't repeat it. We can't go back and say "oh, use this *other* package from Fedora 35 to build this package in Fedora 36".

> Or does "ending up in" still mean it would have to be treated as a prebuilt?

No. It's just an ordinary existing package available in the distro, just like how `cmake` and `clang` are.

Comment (dagood) [reply dir.props]: I'm back to being a little confused trying to understand the full picture. 😕 If we have the Fedora 35 SDK available, I see these as the options for how to build the Fedora 36 SDK:

1. Use a prebuilt SDK (Microsoft or F35, doesn't matter) to build the bootstrap SDK, then use the bootstrap SDK to build the F36 SDK. (Bootstrap flow.)
2. Use the F35 SDK as the N-1 when building the initial F36 SDK. (N-1 "line of continuity" flow.)
   * This is what I interpret "We can't do this in Fedora" being about. But I don't see an explanation why we can't, and your latest comments seem to point fairly strongly towards being able to do this for the first SDK build on F36.

I assume (1) is worse than (2), because you need to leverage the bootstrap exception (risk of carrying prebuilts via a build infra bug) and the extra build makes it take longer.

Comment (omajid) [reply dir.props]: > I assume (1) is worse than (2), because you need to leverage the bootstrap exception (risk of carrying prebuilts via a build infra bug) and the extra build makes it take longer.

Yes, for sure.

> I'm back to being a little confused trying to understand the full picture

Sorry for not being clear about this. Let me try again.

When it's time to create a new version of Fedora, say Fedora N, they start by copying all the packages from Fedora N-1 and putting it a new repository and calling the set Fedora N.

So if Fedora 35 has a .NET Core SDK, it is copied over and Fedora 36 start with that SDK as the base SDK. To put it another way, as soon as Fedora N is created, there is a .NET Core SDK in it. We can then build an updated SDK in Fedora N using that package initially copied over from Fedora N-1.

That allows all distro-packagers to skip bootstrap every packge (unless there are other issues) every time Fedora N is created based off of Fedora N-1.

Other than that, Fedora N, Fedora N-1 and Fedora N+1 have no relationship to each other, and can't use each other's packages.

As an example: once Fedora N is created from Fedora N-1, I can't ask to use an updated package (eg `gcc`) from Fedora N-1 to build something else in Fedora N.

Does that help clarify things?

Comment (dagood) [reply dir.props]: It reinforces it for sure. I guess I was hung up on this:

> > In what situations can you use RHEL 7 bits to build a package for RHEL 8? [...]
> 
> We can't do this in Fedora. [...]

My question was intended to be about this exact scenario, so this seemed like a concise/direct "no", and it's been frying my brain trying to figure out what differs between what we're talking about. 😛 Putting that specific answer aside as some kind of interpretation issue, this is all pretty clear now and matches what I was expecting. Thanks for taking the time to write it all down.

Comment (dagood) [reply dir.props]: Applying this back to the question of the PR... the hypothetical scenario where portable source-built artifacts could solve an issue is if Fedora 36 stopped including libunwind. (Or similar dependency.)

So: that causes the Fedora 35 SDK binaries to no longer work on Fedora 36. **However**, libunwind from Fedora 35 would actually still be available during (only!) the initial build of Fedora 36, so the distro maintainer could actually install Fedora 35's libunwind to successfully use the Fedora 35 SDK to build the initial Fedora 36 SDK.

Does that check out? (We still land in the same place: distro maintainers have no need for carrying along otherwise-unused portable builds in their build chain.)

Comment (omajid) [reply dir.props]: > So: that causes the Fedora 35 SDK binaries to no longer work on Fedora 36.

That's true.

> However, libunwind from Fedora 35 would actually still be available during (only!) the initial build of Fedora 36, so the distro maintainer could actually install Fedora 35's libunwind to successfully use the Fedora 35 SDK to build the initial Fedora 36 SDK.

If a package like `libunwind` is removed, it is generally removed just before Fedora 36 is created. So Fedora 36 would have the initial .NET Core SDK with a broken/missing dependency. Fedora tooling will generally flag something like this a couple of weeks before it happens so developers have time to fix it. ([Here's an example](https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/2QTQOC34OTSFVHZXHKC5D62HURRDDXG3/)). If we somehow miss those and end up with a .NET Core SDK with a missing dependency, I think we would have to go back to bootstrapping.

 > We still land in the same place: distro maintainers have no need for carrying along otherwise-unused portable builds in their build chain.

Agreed.

File: dir.props [hunk 3]
```diff
@@ -10,11 +10,21 @@
...
     <_IsBootstrapping Condition="'$(BootstrapBuildToolsDir)' != ''">true</_IsBootstrapping>
 
-    <PortableBuild Condition="'$(OS)' == 'Windows_NT'">true</PortableBuild>
-    <PortableBuild Condition="'$(PortableBuild)' == ''">false</PortableBuild>
-    <UseSystemLibraries Condition="'$(UseSystemLibraries)' == '' AND '$(PortableBuild)' != 'true'">true</UseSystemLibraries>
-    <UseSystemLibunwind Condition="'$(UseSystemLibunwind)' == ''">$(UseSystemLibraries)</UseSystemLibunwind>
     <UseStableVersions Condition="'$(UseStableVersions)' == ''">true</UseStableVersions>
+
+    <!-- new supported portable/nonportable options, default based on old options if set, otherwise just regular defaults -->
+    <BuildPortableRuntime Condition="'(BuildPortableRuntime)' == '' and '$(PortableBuild)' == 'false'">false</BuildPortableRuntime>
```
Comment (dseefeld) [inline dir.props:16]: Missing $ in the condition on this line and subsequent lines?

File: dir.props [hunk 4]
```diff
@@ -10,11 +10,21 @@
...
+    <BuildPortableRuntime Condition="'(BuildPortableRuntime)' == '' and '$(PortableBuild)' == 'true'">true</BuildPortableRuntime>
+    <BuildPortableRuntime Condition="'(BuildPortableRuntime)' == ''">false</BuildPortableRuntime>
+    <BuildNonPortableRuntime Condition="'(BuildNonPortableRuntime)' == '' and '$(PortableBuild)' == 'false'">true</BuildNonPortableRuntime>
+    <BuildNonPortableRuntime Condition="'(BuildNonPortableRuntime)' == '' and '$(PortableBuild)' == 'true'">false</BuildNonPortableRuntime>
+    <BuildNonPortableRuntime Condition="'(BuildNonPortableRuntime)' == ''">true</BuildNonPortableRuntime>
+    <BuildPortableSdk Condition="'(BuildPortableSdk)' == '' and '$(PortableBuild)' == 'false'">false</BuildPortableSdk>
+    <BuildPortableSdk Condition="'(BuildPortableSdk)' == '' and '$(PortableBuild)' == 'true'">true</BuildPortableSdk>
+    <BuildPortableSdk Condition="'(BuildPortableSdk)' == ''">false</BuildPortableSdk>
```
Comment (dseefeld) [inline dir.props:24]: Is BuildPortableSdk  & BuildNonPortableSdk used anywhere?

File: dir.props (hunk 3 above)
Comment (crummel) [reply dir.props:16]: Yes, thanks.

File: dir.props (hunk 4 above)
Comment (crummel) [reply dir.props:24]: Yeah, the idea is that these are the new control variables for both portable and non-portable builds: https://github.com/dotnet/source-build/pull/1530/files#diff-92e85f9e6b0e59ac672e4c0aeb30cbe5R50.  Then you can either turn these on and off manually, or our legacy variables will default them as before.

**Comment by crummel** [comment]
This was incorporated into update PRs and is no longer needed separately.
